# origin: Debian, #58566
# cset: n/a
# inclusion: not submitted
# description: fix argument handling in main.c
# revision date: 2004-09-03

diff -urN kernel-source-2.4.26/init/main.c kernel-source-2.4.26-1/init/main.c
--- kernel-source-2.4.26/init/main.c	2003-11-29 05:26:21.000000000 +1100
+++ kernel-source-2.4.26-1/init/main.c	2003-11-29 20:53:51.000000000 +1100
@@ -235,11 +235,14 @@
 {
 	char *next,*quote;
 	int args, envs;
+	char *peq;
+	int oenvs, i;
 
 	if (!*line)
 		return;
 	args = 0;
 	envs = 1;	/* TERM is set to 'linux' by default */
+	oenvs = envs;
 	next = line;
 	while ((line = next) != NULL) {
                 quote = strchr(line,'"');
@@ -256,6 +259,8 @@
                 }
                 if (next != NULL)
                         *next++ = 0;
+		if (!*line)
+			continue;
 		if (!strncmp(line,"init=",5)) {
 			line += 5;
 			execute_command = line;
@@ -274,7 +279,20 @@
 		 * Then check if it's an environment variable or
 		 * an option.
 		 */
-		if (strchr(line,'=')) {
+		if ((peq = strchr(line,'=')) != NULL) {
+			/*
+			 * Make sure we don't duplicate what's in the
+			 * initial environment.  We don't check for
+			 * duplicates in line since the user can sort
+			 * that one out.
+			 */
+			for (i = 0; i <= oenvs; i++)
+				if (strncmp(line,envp_init[i],peq-line) == 0) {
+					envp_init[i] = line;
+					break;
+				}
+			if (i <= oenvs)
+				continue;
 			if (envs >= MAX_INIT_ENVS)
 				break;
 			envp_init[++envs] = line;
